docs/PYTHONIC_DESIGN_PRINCIPLES.md

```diff
@@ -0,0 +1,73 @@
+
+While we try to maintain parity between dotnet, python and java for Semantic Kernel, it is also important to allow for language-specific idioms and for best practices to be followed. This decision will help guide the development of the Python version of Semantic Kernel.
+
+## Context
+There are some structural differences between dotnet and python that bear a closer look, as they have a large effect on designing python applications vs. designing dotnet applications. These differences include:
+- Inheritance
+  - in dotnet a class can only inherit from one class, while in python a class can inherit from any number of other classes
+  - in dotnet interfaces play a major role, while in python a interface as such does not exist, they can be replicated with abstract base classes (ABC's) and somewhat with Protocols, but those are not the same, for instance a ABC can have implemented methods, while an interface cannot.
+  - further a regular class can have a abstract method, which is not implemented, making the class act like a ABC without explicitly being one.
+  - python protocols are a way to define a interface, but they are not enforced but used for duck-typing, so that they can be checked at runtime as well as by type checkers, but the developer does not have to worry about them, they can implement their class that adheres to the protocol without explicitly stating that it does, in other words, if it quacks like a duck it is treated as a duck.
+  - a developer can also take any existing class and subclass it and change one or more methods within it, without having to reimplement everything that is in the original class, this is also a quite common pattern in python and one of the reasons that python is so flexible and didn't introduce interfaces (and ABC and Protocols only in later versions).
+- Dependency Injection
+  - In python everything is a object, which includes classes and functions, this means that dependency injection can be done by passing a function as a parameter to another function, or by passing a class as a parameter to another class, and this is a very common pattern in python, there isn't even really a name for it.
+  - There are some packages out there that replicate the behavior of dependency injection, but they are not as widely used, and are not needed for the current codebase.
+- Typing
+  - While python is untyped, it is a best practice to type everything and we do, and we check those types using mypy.
+- Private vs Public
+  - In python there is no private or protected, everything is public, but there is a convention to use a single underscore to denote a private method or variable, and a double underscore to denote a protected method or variable, but this is only a convention and not enforced by the language.
```

Comment:
In terms of conventions, should we mention we follow the Google Python guide (do we?)?